Initial work towards demonstration code and documentation

d33bs opened the pull request:
This PR adds changes which are intended to demonstrate how to use Python code through Anaconda on [UCB RC Alpine](https://curc.readthedocs.io/en/latest/clusters/alpine/index.html) to a CU Anschutz audience. The majority of the content surrounds explaining what Alpine "is" and how to use the code from this repository to help them learn how to use it effectively for their research purposes.

Outside the scope of this PR and as a heads up just in case it changes the review somehow: I intend to reuse some of this content for a future set-website blog article surrounding the same topic. My feeling here was that someone may stumble across this information within Github directly (in a self-contained way) or may find it more appealing to read through the set-website blog. Generally open to the approach here.

Thanks in advance for any comments and suggestions you may have! Please don't hesitate to let me know if you have any questions.

d33bs committed:
add some example code and env

d33bs committed:
add alpine diags and descriptions, begin slurm sec

d33bs committed:
implementation diagram and init content; run scrpt

d33bs committed:
step 0 access content

d33bs committed:
add data transfer + prepare data on alpine content

d33bs committed:
add comments to run script

d33bs committed:
quotes display tweaks

d33bs committed:
replace fontawesome with emoji symbols

github mermaid rendering does not include fontawesome compatibility.

d33bs committed:
add args

d33bs committed:
Update run_script.sh

d33bs committed:
Update run_script.sh

d33bs committed:
Update run_script.sh

d33bs committed:
add env variable

d33bs committed:
yml file ext

d33bs committed:
remove the environment if it exists

d33bs committed:
yes remove env

d33bs committed:
linting; named variable use in python

d33bs committed:
adding logging for stdout from sbatch run

d33bs committed:
implementation process and modified logging file

d33bs committed:
add globus transfer directions

d33bs committed:
Update README.md

d33bs committed:
spacing

vincerubinetti commented on README.md:
```diff
@@ -1,2 +1,395 @@
+
+- __Compute resources:__ Leveraging otherwise cost-prohibitive amounts of memory, CPU, GPU, etc. for processing data.
+- __Long-running jobs:__ Completing long-running processes which may take hours or days to complete.
+- __Collaborations:__ Sharing a single implementation environment for reproducibility within a group (avoiding "works on my machine").
```
`"works on my machine" inconsistency issues)`.

d33bs replied:
Thanks! A lot more descriptive.

vincerubinetti commented on README.md:
```diff
@@ -1,2 +1,395 @@
+1. __External / remote storage:__ Users are encouraged to explore external data storage options for long-term hosting.<br>
+   Examples may include the following:
+
+   - __[Petalibrary](https://www.colorado.edu/rc/resources/petalibrary)__: subsidized external storage host from University of Colorado Boulder's Research Computing (requires specific arrangements outside of Alpine).
```
`PetaLibrary`

vincerubinetti replied:
Side note of no consequence, you may want to find+replace `GitHub` to the canon capitalization.

d33bs replied:
Great points, thank you, will update these in further commits!

vincerubinetti commented on README.md:
```diff
@@ -1,2 +1,395 @@
+_An example of what this preparation section might look like in your Alpine terminal session._
+
+Next we need to prepare our code within Alpine.
+We do this to balance the fact that we may develop and source control code outside of Alpine and needing to periodically synchronize it with updates.
```
`need`

d33bs replied:
Thanks! I understood this to mean: "the use of the word 'need' is questionable". I updated this as follows:
```text
Next we will prepare our code within Alpine.
We do this to balance the fact that we may develop and source control code outside of Alpine.
```

vincerubinetti commented on run_script.sh:
```diff
@@ -0,0 +1,123 @@
+# Slurm directives:
+# -------------------
+# Below are configurations for Slurm, letting it know
+# what and how you'd like to use resources on Alpine.
```
`that specify which resources you'd like to use and how you'd like to use them on Alpine`?

d33bs replied:
Thanks - way clearer! I've added this into a further commit.

vincerubinetti commented on run_script.sh:
```diff
@@ -0,0 +1,123 @@
+#SBATCH --job-name=example-hpc-alpine-python
+
+# Tells Slurm to gather standard output from running this
+# file and send to a specific file.
```
`write it to a specific file`

vincerubinetti commented on run_script.sh:
```diff
@@ -0,0 +1,123 @@
+# as a filepath for exporting data which we pass in here
+# from an sbatch exported variable name which is
+# received as a named argparse variable within Python
+# using the same name.
```
Maybe clean up or simplify this paragraph. Probably can rely on other comments and the script itself for finer explanation.

d33bs replied:
Thanks, great point! I've updated this to retain some of the information and simplify. My hope here is that one wouldn't mistake the sbatch environment variable for the python argparse value.

vincerubinetti commented on run_script.sh:
```diff
@@ -0,0 +1,123 @@
+#SBATCH --mail-user=your-email-address-here@cuanschutz.edu
+
+# Indicate which notifications you'd like to receive from Alpine
+# this can also be set to START, END, or FAIL.
```
Missing period? May also want to check the rest of the comments for missing periods and capitalizations.

d33bs replied:
Thanks! Updated this and others in further commits.

vincerubinetti commented on code/example.py:
```diff
@@ -0,0 +1,35 @@
+"""
+An example Python file which creates random data and exports it to a location specified
+by way of a
```
`to a location specified in a command line argument`.

d33bs replied:
Thank you! I've updated to add this in.

vincerubinetti reviewed (approved):
I can't verify the accuracy of the Slurm and Alpine info, but everything here looks good. Cleanly explained and laid out. If I were needing to make use of Alpine, I think I could do it without problems following the instructions here.

d33bs committed:
more descriptive text

Co-Authored-By: Vincent Rubinetti <8326331+vincerubinetti@users.noreply.github.com>

d33bs committed:
appropriate product naming capitalizations

Co-Authored-By: Vincent Rubinetti <8326331+vincerubinetti@users.noreply.github.com>

d33bs committed:
simplifying code preparation description

Co-Authored-By: Vincent Rubinetti <8326331+vincerubinetti@users.noreply.github.com>

d33bs committed:
more descriptive slurm directive documentation

Co-Authored-By: Vincent Rubinetti <8326331+vincerubinetti@users.noreply.github.com>

d33bs committed:
write instead of send

Co-Authored-By: Vincent Rubinetti <8326331+vincerubinetti@users.noreply.github.com>

d33bs committed:
simplify run script python docs

Co-Authored-By: Vincent Rubinetti <8326331+vincerubinetti@users.noreply.github.com>

d33bs committed:
capitalizations and sentence endings

Co-Authored-By: Vincent Rubinetti <8326331+vincerubinetti@users.noreply.github.com>

d33bs committed:
docstring sentence

Co-Authored-By: Vincent Rubinetti <8326331+vincerubinetti@users.noreply.github.com>

falquaddoomi commented on README.md:
```diff
@@ -1,2 +1,395 @@
+
+Using Alpine effectively involves knowing how to leverage Slurm.
+A simplified way to understand how Slurm works is through the following sequence.
+Please note that some steps and additional complexity are obscured for the purposes of providing a basis of understanding.
```
```suggestion
Please note that some steps and additional complexity are omitted for the purposes of providing a basis of understanding.
```

falquaddoomi commented on README.md:
```diff
@@ -1,2 +1,395 @@
+_Diagram showing high-level user workflow and Alpine components._
+
+Alpine's compute resources are used through compute nodes in a system called [Slurm](https://github.com/SchedMD/slurm).
+Slurm helps coordinate shared and configurable access to the compute resources.
```
Feel free to take this as-is, change it, or leave it out, but I think diving in a bit deeper into what Slurm is here would make concepts later in the document easier to explain. Here's my attempt; new terms are bolded for emphasis the around the first time they're defined:

```suggestion
Slurm is a system that a large number of users to run jobs on a cluster of computers; the system figures out how to use all the computers in the cluster to execute all the user's jobs fairly (i.e., giving each user approximately equal time and resources on the cluster). A *job* is a request to run something, e.g. a bash script or a program, along with specifications about how much RAM and CPU it needs, how long it can run, and how it should be executed.

Slurm's role in general is to take in a job (submitted via the `sbatch` command) and put it into a *queue* (also called a "partition" in Slurm). For each job in the queue, Slurm constantly tries to find a computer in the cluster with enough resources to run that job, then when an available computer is found runs the program the job specifies on that computer. As the program runs, Slurm records its output to files and finally reports the program's exit status (either completed or failed) back to the job manager.

Importantly, jobs can either be marked as *interactive* or *batch*. When you submit an interactive job, `sbatch` will pause while waiting for the job to start and then connect you to the program, so you can see its output and enter commands in real time. On the other hand, a *batch* job will return immediately; you can see the progress of your job using `squeue`, and you can typically see the output of the job in the folder from which you ran `sbatch` unless you specify otherwise.
```

d33bs replied:
Cheers, this is fantastic! I'll commit this in.

falquaddoomi commented on README.md:
```diff
@@ -1,2 +1,395 @@
+> ℹ️ __Wait, what are "nodes"?__
+>
+> A simplified way to understand the architecture of Slurm on Alpine is through login and compute "nodes" (computers).
+> Login nodes act as a way to prepare and submit processes which will be completed on compute nodes.
```
You might want to call back to the definition of a job from before, e.g.:

```suggestion
> Login nodes act as a place to prepare and submit jobs which will be completed on compute nodes. Login nodes are never used to execute Slurm jobs, whereas compute nodes are exclusively accessed via a job.
```

d33bs replied:
Thanks - way clearer! Committing this in.

falquaddoomi commented on README.md:
```diff
@@ -1,2 +1,395 @@
+> Login nodes have limited resource access and are not recommended for running procedures.
+
+One can interact with Slurm on Alpine by use of [Slurm interfaces and directives](https://curc.readthedocs.io/en/latest/clusters/alpine/examples.html).
+A quick way of accessing Alpine resources is through the use of the `acompile` command, which references a script with common Slurm configurations.
```
Again, a callback to interactive jobs:

```suggestion
A quick way of accessing Alpine resources is through the use of the `acompile` command, which starts an interactive job on a compute node with some typical default parameters for the job. Since `acompile` requests very modest resources (1 hour and 1 CPU core at the time of writing), you'll typically quickly be connected to a compute node. For more intensive or long-lived interactive jobs, consider using `sinteractive`, which allows for more customization [Interactive Jobs](https://curc.readthedocs.io/en/latest/running-jobs/interactive-jobs.html).
```

d33bs replied:
Thanks, this is a great expansion, committing this in as well.

falquaddoomi commented on README.md:
```diff
@@ -1,2 +1,395 @@
+These may vary contingent on the external data storage being referenced, the code involved, or your group's available resources.
+Please reference the following documentation from the University of Colorado Boulder's Research Computing regarding data transfers: [The Compute Environment - Data Transfer](https://curc.readthedocs.io/en/latest/compute/data-transfer.html).
+__Please note:__ due to the authentication configuration of Alpine many local or SSH-key based methods are not available for CU Anschutz users.
+As a result, [Globus](https://www.globus.org/) represents one of the best options available (see [3. 📂 Transfer data results](#3-%F0%9F%93%82-transfer-data-results) below).
```
Awesome; AFAIK Globus is (currently) the recommended mechanism for getting data in and out of Alpine, and it's great to see your detailed tutorial on it later in the text. It may be worth making special note here that Globus can be used both to transfer data into Alpine and out of it, e.g.:

```suggestion
As a result, [Globus](https://www.globus.org/) represents one of the best options available (see [3. 📂 Transfer data results](#3-%F0%9F%93%82-transfer-data-results) below). While the Globus tutorial in this document describes how you can download data from Alpine to your computer, note that you can also use Globus to transfer data to Alpine from your computer.
```

d33bs replied:
Thanks, great points! Committing this in as well.

falquaddoomi commented on README.md:
This is probably work for a later date, but I could imagine splitting this README into a number of smaller tutorials on:

- submitting jobs to Slurm
- handling environments via Anaconda
- managing data on Alpine
  - e.g., the home, project, and scratch folders and their intended uses/limitations, and how to get data in and out (via Globus)

and retaining this README as an overview of the system and links to those tutorials.

I could also see keeping the value of it all in one file, too, so up to you, really.

d33bs replied:
Thanks for this - I agree that we should probably have deeper or more specific tutorials on these topics. Candidly I thrashed a bit over whether to include warnings or links to further guidance on those topics to help keep this from getting too long. With this article I sought to create a Pareto principle 20% (of knowledge / training / code) which could offer high value return for the audience's time spent (I don't know if we reached 80%! 🙂 ). At the edges of the content for things which we don't cover my hope is that this at least ignites an eagerness on the audience's part to customize or reach a bit further than we go here.

Other ideas on the content might be that it is used as a short workshop reference to help assist with various issues that could come up along the way; there are a many gaps and questions I can see coming up along the way. Maybe an async way to assist here knowing that schedules could be difficult would be a screenshared recording adjacent to this to help walk through things which is also complemented by the in-depth / focused articles you mentioned.

d33bs replied:
Created #2 to expand here based on what you suggested, resolving here for now.

falquaddoomi reviewed (approved):
Great work! I have no issues with the content itself; I think it's all good and quite valuable. I'm going to go ahead and approve it since we're on a bit of a tight deadline, but feel free to look at my comments and integrate them as you like.

That said, I don't know enough about the target audience quite yet to know what will make sense to them, what will go over their heads, and what they'll want to do that might be missing. IMHO, I think we should trial it with some people and keep this as a living document that we simplify or expand according to our test readers' feedback.

d33bs committed:
Apply suggestions from code review

Co-authored-by: Faisal Alquaddoomi <falquaddoomi@gmail.com>

d33bs committed:
linting

d33bs committed:
add authorship team / department links

d33bs commented:
Thanks @vincerubinetti and @falquaddoomi for the great reviews and content suggestions! I think I've addressed all comments and will be merging this in.
